File: test-automation-core/src/main/java/com/nortal/test/core/rest/interceptors/ReportInterceptor.java
package com.nortal.test.core.rest.interceptors;

import java.io.IOException;
import org.springframework.stereotype.Component;
import com.nortal.test.core.services.report.cucumber.RequestResponseReportFormatter;
import lombok.RequiredArgsConstructor;
import okhttp3.Interceptor;
import okhttp3.Response;

@Component
@RequiredArgsConstructor
public class ReportInterceptor implements Interceptor {

	private final RequestResponseReportFormatter formatter;

	@Override
	public Response intercept(final Interceptor.Chain chain) throws IOException {

		try {
			final Response response = chain.proceed(chain.request());
			formatter.formatAndAddToReport(response.request(), response);
			return response;
		} catch (Exception e) {
			formatter.formatAndAddToReport(chain.request());
			throw e;
		}
	}

}

File: test-automation-core-old/src/main/java/com/nortal/test/core/rest/interceptors/ReportInterceptor.kt
package com.nortal.test.core.rest.interceptors

import com.nortal.test.core.services.report.cucumber.RequestResponseReportFormatter
import com.nortal.test.core.services.report.cucumber.RequestResponseReportFormatter.formatAndAddToReport
import lombok.RequiredArgsConstructor
import okhttp3.Interceptor
import okhttp3.Response
import org.springframework.stereotype.Component
import java.io.IOException
import java.lang.Exception

@Component
@RequiredArgsConstructor
class ReportInterceptor : Interceptor {
    private val formatter: RequestResponseReportFormatter? = null
    @Throws(IOException::class)
    override fun intercept(chain: Interceptor.Chain): Response {
        return try {
            val response = chain.proceed(chain.request())
            formatter!!.formatAndAddToReport(response.request(), response)
            response
        } catch (e: Exception) {
            formatter!!.formatAndAddToReport(chain.request())
            throw e
        }
    }
}